example/black_scholes.cpp

```diff
@@ -44,10 +44,11 @@ promote<Price,Sigma,Tau,Rate>
 int main()
 {
   const double K = 100.0; // Strike price.
-  const auto S     = make_fvar<double,3>(105); // Stock price.
-  const auto sigma = make_fvar<double,0,3>(5); // Volatility.
-  const auto tau   = make_fvar<double,0,0,1>(30.0/365); // Time to expiration in years. (30 days).
-  const auto r     = make_fvar<double,0,0,0,1>(1.25/100); // Interest rate.
+  const auto variables = make_ftuple<double,3,3,1,1>(105, 5, 30.0/365, 1.25/100);
+  const auto& S     = std::get<0>(variables); // Stock price.
+  const auto& sigma = std::get<1>(variables); // Volatility.
+  const auto& tau   = std::get<2>(variables); // Time to expiration in years. (30 days).
+  const auto& r     = std::get<3>(variables); // Interest rate.
```

Comment:
Is there any reason these now need to be references? It seems intuitive to me that fvars should be able to be copied by value cheaply (unless the number of taylor expansion terms is huge I guess). The reason I ask is that I wonder if this might be written more concisely using structured bindings along the lines of
```auto [S, sigma, tau, r] = make_ftuple<double, 3,3,1,1>(105, 5, 30.0/365, 1.25/100);```
Not sure how you'd incorporate the comments.

Reply:
> Is there any reason these now need to be references?

They don't need to be; I just thought it would be useful to see that as a possibility.

> It seems intuitive to me that fvars should be able to be copied by value cheaply (unless the number of taylor expansion terms is huge I guess). The reason I ask is that I wonder if this might be written more concisely using structured bindings along the lines of
`auto [S, sigma, tau, r] = make_ftuple<double, 3,3,1,1>(105, 5, 30.0/365, 1.25/100);`

That can work just as well - perhaps I'll update the example accordingly.

The use of `std::apply` in `example/mixed_partials.cpp` may be one reason why one might want to use the `tuple` object directly.

Reply:
You're right. I'm not questioning the usefulness of storing the variables in a tuple. Rather, when you do intend to put them in a bunch of variables (and don't still need them in the tuple), then the option exists to do so concisely.

In the case where you want them in both forms I wonder if it would be clearer/briefer to use make_tuple again on the variables. Is there any kind of assertion on use that they're ordered correctly? There is a single correct ordering inherent in their type right?

Reply:
> In the case where you want them in both forms I wonder if it would be clearer/briefer to use make_tuple again on the variables.

That can certainly be done. However the main motivation of `make_ftuple` is the convenience of declaring multiple independent variables at once without having to keep track of the different dimensions of each independent variables. All that is needed are two lists:
1. Maximum order of derivative of x_i to be calculated.
2. Value of x_i.

List (1) is given in the `make_ftuple` function template parameter list, and list (2) is given in the function parameters.

> Is there any kind of assertion on use that they're ordered correctly?

Lists (1) and (2) must be in the same order. (I don't see how one could possibly include a code-level assertion to enforce this in case that is what you are asking.)

> There is a single correct ordering inherent in their type right?

Not sure what you mean by this. The function parameters, e.g. (105, 5, 30.0/365, 1.25/100), are cast to the type given as the first function template parameter (`double`).

Reply:
I meant if you call ```make_ftuple<T,1,1,1>``` you effectively get a ```tuple<fvar<T,1>, fvar<T,0,1>, fvar<T,0,0,1>>```
if you were to use structured bindings you you end up with 3 variables a,b,c corresponding to these types. If you were to subsequently require a tuple I was saying that rather than keeping the original it might make sense to just make a new one. My question is is there a way to prevent you making/using a ```tuple<fvar<T,0,1>, fvar<T,0,0,1>, fvar<T,1>> for example. However in retrospect I'm not even sure that this would cause any problems. I guess it depends what you do with the tuple and right now we're not doing anything with it.

Reply:
> in retrospect I'm not even sure that this would cause any problems

No it wouldn't, as long as you query the derivatives in the correct order. E.g. if `auto y = std::apply(f, tuple);` then `y.derivative(0,0,1)` gives the derivative w.r.t. `c` regardless of the order of `tuple` and `f`'s parameters, as long as they are consistent with each other.